DIPs/DIP-1NN4-QFS.md

```diff
@@ -0,0 +1,573 @@
+* Only on a functional's parameter type, warrant attributes give rise to a *requirement* that the functional *needs* to work properly.
+
+As an example, consider a functional taking a `pure` delegate parameter.
+The functional might make use of memoization or the fact that the parameter's return values are
+[unique](https://dlang.org/spec/const3.html#implicit_qualifier_conversions) in its internal logic.
+In the case of uniqueness, omitting the requirement will result in illegal code, since return values of impure functions generally cannot assumed to be unique.
+In the case of memoization, omitting the requirement will not result in a compile error, but unexpected behavior, when used improperly.
+In most cases, however, the requirement is merely to match the functionals guarantee.
+
+In the current state of the language,
+a funcional cannot have strong guarantees and weak requirements at the same time.
+Most programmers opt against warrant attributes, i.e. for weak requirements
+and therefore needlessly weaken the guarantees.
+
+The DIP helps this situation by giving calls to parameters a free pass
+when checking a functional for satisfying the conditions of a warrant attribute.
+
+At the point where the functional is called, the type system has all the necessary information available
+to determine if the execution of it will comply with the warrant attributes of the context.
+
+This entails that not all calls to a functional annotated with a warrant attribute will result in call expressions
+that is considered in compliance with that warrant attribute.
+This is, however, less of a problem than it seems at first glance:
+* Consider a `@system` or `@trusted` context calling a `@safe` functional with a `@system` argument.
+The context does not expect an execution satisfying the formal `@safe` constraints, therefore
+the fact that *the call* to the `@safe` annotated function is not formally considered `@safe` can be mostly ignored.
+* Consider a `@safe` context calling a `@safe` functional with a `@system` argument.
+This is illegal, since a `@safe` functional makes guarantees for its internal logic.
+What the `@system` callback does, is among the concerns of the context,
+and in a `@safe` context, calling a  `@system` function is illegal.
+
+Especially in meta-programming, it might not be clear at all whether a callback's type has a warrant attribute or not.
+For `@safe` and `@nogc`, this is unproblematic, since these are only interesting from a safety or resource perspective,
+but no program's logic depends on the guarantees these attributes make: A memory unsafe program is broken in itself and
+GC allocating may at worst slow down a program unexpectedly due to the GC issuing a collection cycle.
+(Note that `@nogc` alone does not guarantee that no dynamic allocation using other ways take place.)
+[Author's note: I'm not completely sure. Please have a think whether these claims are really true.]
```

Comment:
Yeah, `malloc` can be used, but that's tangential and I'll just remove the mention.

Reply:
Done.